flutter-studio/src/io/flutter/module/FlutterModuleModel.java

```diff
@@ -5,27 +5,215 @@
+import com.intellij.openapi.util.io.FileUtilRt;
+import com.intellij.openapi.vfs.VfsUtil;
+import com.intellij.openapi.vfs.VirtualFile;
+import io.flutter.FlutterUtils;
+import io.flutter.actions.FlutterBuildActionGroup;
 import io.flutter.project.FlutterProjectCreator;
 import io.flutter.project.FlutterProjectModel;
+import io.flutter.pub.PubRoot;
+import io.flutter.sdk.FlutterSdk;
+import io.flutter.utils.AndroidUtils;
+import java.io.File;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
 import org.jetbrains.annotations.NotNull;
 
 public class FlutterModuleModel extends FlutterProjectModel {
   public FlutterModuleModel(@NotNull FlutterProjectType type) {
     super(type);
   }
 
+  @Override
+  public boolean shouldOpenNewWindow() {
+    return !isModule();
+  }
+
   @Override
   protected void handleFinished() {
     // Do not call the superclass method.
-    if (projectType().get().isPresent() && projectType().get().get() == FlutterProjectType.IMPORT) {
-      String location = projectLocation().get();
-      assert (!location.isEmpty());
+    if (isModule()) {
+      // The FlutterProjectType.MODULE is used in two places. In FlutterProjectModel it is used to create a new top-level
+      // Flutter project. Here, it is used to create an Android module for add-to-app. In both cases the Flutter tool creates
+      // a new project. In this case, we go on to link that project into an Android project as a new module.
+      useAndroidX().set(true);
+      // The host project is an Android app. This module should be created in its root directory.
+      // Android Studio supports a colon-separated convention to specify sub-directories, which is not yet supported here.
+      Project hostProject = project().getValue();
+      String hostPath = hostProject.getBasePath();
+      if (hostPath == null) {
+        throw new InvalidDataException(); // Can't happen
+      }
+      projectLocation().set(hostPath);
+      // Create the Flutter module as if it were a normal project that just happens to be located in an Android project directory.
+      new FlutterProjectCreator(this).createModule();
+      // Import the Flutter module into the Android project as if it had been created without any Android dependencies.
       new FlutterModuleImporter(this).importModule();
+      // Link the new module into the Gradle project, which also enables co-editing.
+      new FlutterGradleLinker(this).linkNewModule();
     }
     else {
-      assert (!projectName().get().isEmpty());
-      assert (!flutterSdk().get().isEmpty());
-      new FlutterProjectCreator(this).createModule();
+      if (projectType().get().isPresent() && projectType().get().get() == FlutterProjectType.IMPORT) {
+        String location = projectLocation().get();
+        assert (!location.isEmpty());
```

Comment:
Do we want to assert here? If the location is empty, that will throw an exception.

I don't see where the `location` local variable is used.

Reply:
It's been a long time since I wrote that and don't remember why I wanted that assert. I think it is last-chance error checking since without a location we can't do an import. But the wizard verification code shouldn't let the user get to this point without a valid location.